lib/Sema/TypeCheckError.cpp

```diff
@@ -1505,25 +1552,45 @@ class CheckErrorCoverage : public ErrorHandlingWalker<CheckErrorCoverage> {
   }
 
   ShouldRecurse_t checkThrow(ThrowStmt *S) {
-    checkThrowSite(S, /*requiresTry*/ false,
-                   Classification::forThrow(PotentialReason::forThrow()));
+    checkThrowAsyncSite(S, /*requiresTry*/ false,
+                        Classification::forThrow(PotentialReason::forThrow(),
+                                                 /*async*/false));
     return ShouldRecurse;
   }
 
-  void checkThrowSite(ASTNode E, bool requiresTry,
-                      const Classification &classification) {
+  void checkThrowAsyncSite(ASTNode E, bool requiresTry,
+                           const Classification &classification) {
     MaxThrowingKind = std::max(MaxThrowingKind, classification.getResult());
 
-    switch (classification.getResult()) {
-    // Completely ignores sites that don't throw.
-    case ThrowingKind::None:
-      return;
-
     // Suppress all diagnostics when there's an un-analyzable throw site.
-    case ThrowingKind::Invalid:
+    if (classification.isInvalid()) {
       Flags.set(ContextFlags::HasAnyThrowSite);
       if (requiresTry) Flags.set(ContextFlags::HasTryThrowSite);
       return;
+    }
+
+    // If the call to the function is async, handle it.
+    if (classification.isAsync()) {
+      // Remember that we've seen an async call.
+      Flags.set(ContextFlags::HasAnyAsyncSite);
+      
+      // Diagnose async calls that are outside of an await context.
+      if (!Flags.has(ContextFlags::IsAsyncCovered)) {
+        SourceRange highlight;
+        
+        // Generate more specific messages in some cases.
+        if (auto e = dyn_cast_or_null<ApplyExpr>(E.dyn_cast<Expr*>()))
```

Comment:
Nit: Not important, but question is if could this be a place to use `getAsExpr<ApplyExpr>(E)`?

Reply:
Probably. I don't think that function existed when this was originally written ;)

Reply:
@LucianoPAlmeida Unfortunately it cannot be used here as it's defined in ConstraintSystem.

Reply:
Ah true @theblixguy, so maybe move to a more general header file (as you suggested) would be nice to allow use in other places :)